_posts/2022-05-18-My-OTP-25-highlights.md

```diff
@@ -0,0 +1,713 @@
+for example:
+
+```erlang
+commit_write(OpaqueData) ->
+    B = OpaqueData,
+    case disk_log:sync(B#backup.file_desc) of
+        ok ->
+            case disk_log:close(B#backup.file_desc) of
+                ok ->
+                    case file:rename(B#backup.tmp_file, B#backup.file) of
+                        ok ->
+                            {ok, B#backup.file};
+                        {error, Reason} ->
+                            {error, Reason}
+                    end;
+                {error, Reason} ->
+                    {error, Reason}
+            end;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+```
+
+The code is nested to the extent that shorter aliases must be introduced
+for variables (`OpaqueData` renamed to `B`), and half of the code just
+transparently returns the exact values each function was given.
+
+By comparison, the same code could be written as follows with the new
+construct:
+
+```erlang
+commit_write(OpaqueData) ->
+    maybe
+        ok ?= disk_log:sync(OpaqueData#backup.file_desc),
+        ok ?= disk_log:close(OpaqueData#backup.file_desc),
+        ok ?= file:rename(OpaqueData#backup.tmp_file, OpaqueData#backup.file),
+        {ok, OpaqueData#backup.file}
+    end.
+```
+
+Or, to protect against `disk_log` calls returning something else than `ok |
+{error, Reason}`, the following form could be used:
+
+```erlang
+commit_write(OpaqueData) ->
+    maybe
+        ok ?= disk_log:sync(OpaqueData#backup.file_desc),
+        ok ?= disk_log:close(OpaqueData#backup.file_desc),
+        ok ?= file:rename(OpaqueData#backup.tmp_file, OpaqueData#backup.file),
+        {ok, OpaqueData#backup.file}
+    else
+        {error, Reason} -> {error, Reason}
+    end.
+```
+
+The semantics of these calls are identical, except that it is now
+much easier to focus on the flow of individual operations and either
+success or error paths.
+
+# Compiler news
+* Add compile attribute `-nifs()` to empower compiler and loader with information     about   which functions may be overridden as NIFs by `erlang:load_nif/2`.
+* Improved and more detailed error messages when binary construction with the binary syntax fails. This applies both for error messages in the shell and for `erl_error:format_exception/3,4`.
+
+
+# Crypto and OpenSSL 3.0
+Text to be included.....
+
+# Dialyzer
+
+Optimize operations in the erl_types module. Parallelize the Dialyzer pass remote.
+Added the missing_return and extra_return options to raise warnings when specifications differ from inferred types. These are similar to, but not quite as verbose as overspecs and underspecs.
+Dialyzer now better understands the types for min/2, max/2, and erlang:raise/3. Because of that, Dialyzer can potentially generate new warnings. In particular, functions that use erlang:raise/3 could now need a spec with a no_return() return type to avoid an unwanted warning.
```

Comment:
```suggestion
Dialyzer now supports the `missing_return` and `extra_return` options to raise warnings when specifications differ from inferred types. These are similar to, but not quite as verbose, as `overspecs` and `underspecs`.

Dialyzer now better understands the types for `min/2`, `max/2`, and `erlang:raise/3`. Because of that, Dialyzer can potentially generate new warnings. In particular, functions that use `erlang:raise/3` could now need a spec with a `no_return()` return type to avoid an unwanted warning.
```